rfd/0048-war-dialler-node-tracker.md

```diff
@@ -0,0 +1,199 @@
+---
+authors: Naji Obeid (naji@goteleport.com), David Boslee (david@goteleport.com)
+state: draft
+---
+
+# RFD 48 - Node Tracker Service
+
+## What
+
+This is part one of a two part RFD that aims to improve node-to-proxy connectivity by allowing node agents to connect to a single proxy and be reachable through any other proxy in the same cluster.
+
+This RFD proposes a new service that tracks node-to-proxy relationships, in other terms which nodes are connected to which proxies.
+The second RFD proposes a new proxy-to-proxy tunneling system, and can be tracked [here](insert rfd link here)
+
+## Why
+
+The current behavior of node agents dialing over a reverse tunnel are required to connect to every proxy instance in the cluster. This allows a client to connect to a node through any proxy but it also causes other problems when running proxies behind a load balancer like in our cloud environment. These problems include:
+
+- Ephemeral port exhaustion between a NAT gateway and load balancer. This limits the number of nodes that can be connected behind a single NAT gateway. As more proxies are added fewer nodes are able to connect.
+- Thundering herd when adding, removing, or restarting a proxy. Node agents retry connecting until they randomly get balanced to the desired proxy. The more node agents connected the worse this problem becomes.
+
+Both these issues are mitigated by changing the node agent behavior to connect to a single proxy. Ephemeral port exhaustion is no longer tied to the number of proxies and node agents no longer need to retry until they connect to a specific proxy.
+
+## Details
+
+Here's a small example comparing how proxies would interact with the node tracker:
+
+```
+	Original behaviour:
+                       ┌──────┐
+   ┌───connection─────>│proxy1│<──reverse-tunnel──┐
+   │                   └──────┘                   │
+┌──┴───┐                                        ┌─┴──┐
+│client│                                        │node│
+└──┬───┘                                        └─┬──┘
+   │                   ┌──────┐                   │
+   └───connection─────>│proxy2│<──reverse-tunnel──┘
+                       └──────┘
+	A client can reach the requested node through any proxy because the node has a reverse tunnel open to all proxies in a cluster.
+
+
+
+    New behaviour:
+                         ┌──────┐         ┌────────────┐
+   ┌─(1)─connection-1───>│proxy1├──(2)───>│node tracker│
```

Comment:
what happens if the datacenter with the proxy1 and node-tracker is down/under attack?

Reply:
Can the node tracker be a part of proxy's process (so we remove extra hop - 2)? Then each proxy can have embedded node-tracker process?

Reply:
What protocol client uses to discover all possible proxies in the system? There has to be some reconciliation logic by client to try other endpoints if this one is down.

Reply:
> what happens if the datacenter with the proxy1 and node-tracker is down/under attack?

As you mentioned in another comment down below having a secondary node tracker would mitigate for this. That's not too complex to implement in it's current form, but we will have to look into node tracker being discovered by the proxies rather than having them being statically added to the config of each proxy. 

> Can the node tracker be a part of proxy's process (so we remove extra hop - 2)? Then each proxy can have embedded node-tracker process?

We've briefly looked at implementing consensus algorithms and sharding at the proxy level but we think it's' too early for this project. Taking raft as an example, techinically we could not remove that extra hop as reading/writing should go through the leader, one of the proxies in our case, but we could cheat for reads and tap directly into each proxy's internal database... we might need to look for off the shelf libraries/solutions if that's what we're looking for.

> What protocol client uses to discover all possible proxies in the system? There has to be some reconciliation logic by client to try other endpoints if this one is down.

In my understanding nothing changes from the existing behaviour implemented currently in teleport clients. This change should be transparent to all clients.

Reply:
FWIW this kind of peer to peer system doesn't need a consensus algorithm. There is nothing to come to consensus about really. Each proxy is the single source of truth for its node set. These are totally independent states (i.e. there are no rules that would make one set invalid based on the content of another, and no relevant meaning to ordering across sets). If the proxy is alive and reachable, it's peers can stream join/leave events from it directly. If it isn't, knowing it's state is irrelevant since it can't be dialed.

Reply:
I think it's necessary if we want to embed the node tracker in each proxy and have them have a complete world view of where each node is connected to. Without a consensus algorithm we wouldn't know where to look if a proxy has missing or outdated info except having to possibly dial all the proxies in the cluster to fetch the missing connection.

Reply:
I might not be explaining myself well.

Lets look at an example of a scenario that clearly requires a consensus algorithm: say you have a distributed key-value store with two nodes N1 and N2, and two clients C1 and C2.  C1 performs the operation `SET foo -> bar`, and C2 performs the operation `SET foo -> bin`.  The final value stored at `foo` depends on the ordering of these operations, so we need to have N1 and N2 agree which order to use, or their states become permanently out of sync.

Now lets take a master/replica example.  Say I have a key-value store with master M and replica R.  In this case, M gets to decide which of the two `SET` operations happen first, then it simply *informs* R of the ordering.  The two are in sync because R builds its state based on what M tells it to do.  Easy-peasy.

My point is that the proxy routing system is actually a collection of master/replica relationships, rather than a single distributed state.  Each proxy needs to replicate the connection mapping of its peers, and is the master of its own connection mapping.  Modification of different proxies' connection mappings are independent operations.  We don't need relative ordering between them.  If a node appears to be connected to two proxies, thats isn't a conflict.  We can just try dialing both of them, or select one at random.  It really doesn't matter.

With this in mind, we can just extend a proxy's API to look something like this (pseudocode):

```golang
type EventKind int

const (
    Join EventKind = 0
    Leave EventKind = 1
)

type Event struct {
    Kind EventKind
    NodeID string
}

type Peer interface {
    Dial(DialParams) Conn
    WatchNodes() <-chan Event
}
```

Simple enough to emit events when a node connects/disconnects, and maintain a mapping for each peer based on its join/leave events.  In fact, its even trivial to have checkpointing if we lose connectivity to a given proxy intermittently.  Because each proxy is a master of its own state, it can just associate a nonce with each event, and keep a small backlog so that short reconnects don't require re-replication.

Huge bonus of a system like this: we don't have to ever perform cache evictions.  No need to fuss with TTLs and the like.  If a proxy is unreachable, its state doesn't matter.  If it is reachable, it can inform us in real time if a node leaves.

Reply:
I like it! But I still feel there should still be some sort of governing body involved at some stage, like validating the replicas a proxy has with it's peers for example. Some proxies can fail to receive some mappings which will make trying to connect to a node intermittent based on if you were load balanced to a proxy that knows how to get there or not. In contrast, the tracker solution or a distributed state doesn't suffer from this. 
Both have pros and cons, we'll have to iterate on this to find what works best.